Index: sqlite3/src/whereexpr.c
==================================================================
--- sqlite3/src/whereexpr.c
+++ sqlite3/src/whereexpr.c
@@ -1518,10 +1518,16 @@
     if( ExprHasProperty(p, EP_VarSelect) ) pMaskSet->bVarSelect = 1;
     mask |= exprSelectUsage(pMaskSet, p->x.pSelect);
   }else if( p->x.pList ){
     mask |= sqlite3WhereExprListUsage(pMaskSet, p->x.pList);
   }
+#ifndef SQLITE_OMIT_WINDOWFUNC
+  if( p->op==TK_FUNCTION && p->y.pWin ){
+    mask |= sqlite3WhereExprListUsage(pMaskSet, p->y.pWin->pPartition);
+    mask |= sqlite3WhereExprListUsage(pMaskSet, p->y.pWin->pOrderBy);
+  }
+#endif
   return mask;
 }
 Bitmask sqlite3WhereExprUsage(WhereMaskSet *pMaskSet, Expr *p){
   return p ? sqlite3WhereExprUsageNN(pMaskSet,p) : 0;
 }

Index: test/window1.test
==================================================================
--- sqlite3/test/window1.test
+++ sqlite3/test/window1.test
@@ -717,8 +717,35 @@
 do_execsql_test 17.3 {
   SELECT 10+sum(a) OVER (ORDER BY a) 
   FROM t8 
   ORDER BY 10+sum(a) OVER (ORDER BY a) DESC;
 } {16 13 11}
+
+#-------------------------------------------------------------------------
+#
+reset_db
+do_execsql_test 18.0 {
+  CREATE TABLE t1 ( t1_id INTEGER PRIMARY KEY );
+  CREATE TABLE t2 ( t2_id INTEGER PRIMARY KEY );
+  CREATE TABLE t3 ( t3_id INTEGER PRIMARY KEY );
+
+  INSERT INTO t1 VALUES(1),  (3), (5);
+  INSERT INTO t2 VALUES      (3), (5);
+  INSERT INTO t3 VALUES(10), (11), (12);
+}
+
+do_execsql_test 18.1 {
+  SELECT t1.* FROM t1, t2 WHERE 
+    t1_id=t2_id AND t1_id IN (
+        SELECT t1_id + row_number() OVER ( ORDER BY t1_id ) FROM t3
+    )
+}
+
+do_execsql_test 18.2 {
+  SELECT t1.* FROM t1, t2 WHERE 
+    t1_id=t2_id AND t1_id IN (
+        SELECT         row_number() OVER ( ORDER BY t1_id ) FROM t3
+    )
+} {3}
 
 
 finish_test

